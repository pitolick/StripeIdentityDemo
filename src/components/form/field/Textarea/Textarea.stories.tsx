import type { Meta, StoryObj } from '@storybook/react'
import { Textarea } from './Textarea'

const meta: Meta<typeof Textarea> = {
  title: 'Form/Field/Textarea/Base',
  component: Textarea,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Textarea>

/**
 * 基本形
 */
export const Base: Story = {}

/**
 * プレースホルダー
 */
export const Placeholder: Story = {
  args: {
    placeholder: '入力してください',
  },
}

/**
 * 初期値
 */
export const InitialValue: Story = {
  args: {
    defaultValue: '初期値',
  },
}

/**
 * 行数
 */
export const Rows: Story = {
  args: {
    rows: 5,
  },
}
