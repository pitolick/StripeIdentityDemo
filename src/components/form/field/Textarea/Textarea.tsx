/**
 * テキストエリア非盛業コンポーネント
 * @module Textarea
 */
import React, { forwardRef } from 'react'

/**
 * テキストエリアコンポーネントのプロパティ
 * @typedef {Object} TextareaProps
 * @property {React.TextareaHTMLAttributes<HTMLTextAreaElement>} - テキストエリアタグのHTML属性
 */
export interface TextareaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {}

/**
 * テキストエリア非制御コンポーネント
 * @param {TextareaProps} props - テキストエリアコンポーネントのプロパティ
 * @returns {JSX.Element} - テキストエリアコンポーネント
 */
export const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>((props, ref) => {
  const { className = undefined, ...textareaProps } = props
  const classNameProps = className ? className : ''
  return (
    <>
      <textarea ref={ref} {...textareaProps} className={`${classNameProps} block w-full rounded-md border border-gray-500 px-3 py-2 text-sm md:px-5  md:py-3 md:text-base`} />
    </>
  )
})

Textarea.displayName = 'Textarea'
