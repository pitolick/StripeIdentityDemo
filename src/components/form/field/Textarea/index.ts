export * from './Textarea'
export * from './TextareaControl'
