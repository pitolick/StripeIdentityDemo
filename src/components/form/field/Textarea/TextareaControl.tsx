/**
 * テキストエリア制御コンポーネント（react-hook-form）
 * @module TextareaControl
 */
import React, { forwardRef } from 'react'
import { Textarea, TextareaProps } from './Textarea'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * テキストエリアコンポーネントのプロパティ
 * @memo バリデーションスキーマ（zod）は各入力子コンポーネント内でexportしformコンポーネント内でresolverに指定する
 * @typedef {Object} TextareaControlProps
 */
export interface TextareaControlProps extends TextareaProps, Omit<InputBoxProps, 'children'> {}

/**
 * テキストエリア制御コンポーネント（react-hook-form）
 * @param {TextareaControlProps} props - テキストエリアコンポーネントのプロパティ
 * @returns {JSX.Element} - テキストエリアコンポーネント
 */
export const TextareaControl = forwardRef<HTMLTextAreaElement, TextareaControlProps>((props, ref) => {
  const { nameId, label, note, required = undefined, className, error, contentWidth, ...textareaProps } = props
  const classNameProps = className ? className : ''
  // エラー時のclass名
  const errorClass = error ? `bg-red-pale outline outline-red outline-2 -outline-offset-2` : ''

  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      {/* テキストエリア */}
      <Textarea id={nameId} ref={ref} {...textareaProps} className={`${errorClass} ${classNameProps}`} />
    </InputBox>
  )
})

TextareaControl.displayName = 'TextareaControl'
