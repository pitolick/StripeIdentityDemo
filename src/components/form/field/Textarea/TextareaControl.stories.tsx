import type { Meta, StoryObj } from '@storybook/react'
import { TextareaControl } from './TextareaControl'

const meta: Meta<typeof TextareaControl> = {
  title: 'Form/Field/Textarea/Control',
  component: TextareaControl,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof TextareaControl>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    nameId: 'base',
  },
}

/**
 * プレースホルダー
 */
export const Placeholder: Story = {
  args: {
    placeholder: '入力してください',
    nameId: 'placeholder',
  },
}

/**
 * 初期値
 */
export const InitialValue: Story = {
  args: {
    defaultValue: '初期値',
    nameId: 'initial-value',
  },
}

/**
 * 行数
 */
export const Rows: Story = {
  args: {
    rows: 5,
    nameId: 'rows',
  },
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    label: 'ラベル',
    nameId: 'label',
  },
}

/**
 * 注釈
 */
export const Note: Story = {
  args: {
    note: { list: ['注釈'] },
    nameId: 'note',
  },
}

/**
 * バリデーション：必須
 */
export const Required: Story = {
  args: {
    required: true,
    nameId: 'required',
    label: '必須',
  },
}

/**
 * バリデーション：任意
 */
export const Optional: Story = {
  args: {
    required: false,
    nameId: 'optional',
    label: '任意',
  },
}

/**
 * バリデーション：エラー
 */
export const Error: Story = {
  args: {
    required: true,
    nameId: 'error',
    label: 'エラー',
    error: { type: 'required', message: '必須項目です' },
  },
}
