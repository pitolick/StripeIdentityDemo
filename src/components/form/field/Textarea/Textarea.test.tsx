import React from 'react'
import { render } from '@testing-library/react'
import { Textarea } from './Textarea'

describe('Textareaコンポーネント', () => {
  it('正しくレンダリングされていること', () => {
    const { container } = render(<Textarea />)
    expect(container).toMatchSnapshot()
  })

  it('classNameプロパティが指定された場合、正しく適用されること', () => {
    const { container } = render(<Textarea className="text-red" />)
    expect(container).toMatchSnapshot()
  })
})
