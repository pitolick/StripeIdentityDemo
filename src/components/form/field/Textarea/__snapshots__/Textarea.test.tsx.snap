// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textareaコンポーネント classNameプロパティが指定された場合、正しく適用されること 1`] = `
<div>
  <textarea
    class="text-red block w-full rounded-md border border-gray-500 px-3 py-2 text-sm md:px-5  md:py-3 md:text-base"
  />
</div>
`;

exports[`Textareaコンポーネント 正しくレンダリングされていること 1`] = `
<div>
  <textarea
    class=" block w-full rounded-md border border-gray-500 px-3 py-2 text-sm md:px-5  md:py-3 md:text-base"
  />
</div>
`;
